src/plugins/intel_cpu/src/infer_request.cpp

```diff
@@ -135,6 +136,15 @@ void SyncInferRequest::infer() {
 
     push_input_data();
 
+    auto network_memory_control = m_graph->getGraphContext()->getMemoryControl();
+    if (!network_memory_control) {
+        OPENVINO_THROW("Memory control unit is not initilized for graph: ", m_graph->GetName());
+    }
+
+    if (!network_memory_control->allocated()) {
+        network_memory_control->allocateMemory();
+    }
+
```

Comment:
There is no problem to move it into a graph Infer() call, besides it will introduce additional overhead for inner graphs as well, of course negligible.
But let's try to look at this from another perspective.
Who has the ability to release the memory? With the current assumptions it is the user of CompiledModel.
CompiledModel than forwards this to the CPU graph. It would be even better if we would somehow restrict the graph itself to release its own memory, which sounds logical.
So, it is not graph itself who initiates the memory release, so it should not be the graph itself who restores it.
Basically, the requirement can be formulated as: the user of CPU graph must ensure that the memory for the graph is allocated.
So, for example, the nodes with inner graphs never call releaseMemory(), so they should not care about restoring it.

Reply:
> It would be even better if we would somehow restrict the graph itself to release its own memory, which sounds logical.

This solution looks even better, agree. Since in this PR only one memory control object is allowed per compiled network, we could transfer the responsibility of implementing memory release to the main graph, and provide an interface call. But again it brings into discussion the necessity of splitting the graph class into two types: a main graph and a nested graph, as the latter for example mustn't have the ability of releasing memory in the model we're discussing, as the memory control is allowed only for the main graph.